activation/handler.go

```diff
@@ -220,72 +174,176 @@ func (h *Handler) SyntacticallyValidateAtx(ctx context.Context, atx *types.Activ
 		}
 		if d := atx.PubLayerID.Difference(posAtx.PubLayerID); d > h.layersPerEpoch {
-			return nil, fmt.Errorf("expected distance of one epoch (%v layers) from pos atx but found %v", h.layersPerEpoch, d)
+			return nil, fmt.Errorf("expected distance of one epoch (%v layers) from positioning atx but found %v", h.layersPerEpoch, d)
 		}
 		baseTickHeight = posAtx.TickHeight()
-	} else {
-		publicationEpoch := atx.PublishEpoch()
-		if !publicationEpoch.NeedsGoldenPositioningATX() {
-			return nil, fmt.Errorf("golden atx used for atx in epoch %d, but is only valid in epoch 1", publicationEpoch)
-		}
 	}
 
 	expectedChallengeHash, err := atx.NIPostChallenge.Hash()
 	if err != nil {
-		return nil, fmt.Errorf("failed to compute NIPost's expected challenge hash: %v", err)
+		return nil, fmt.Errorf("failed to compute NIPost's expected challenge hash: %w", err)
 	}
 
 	h.log.WithContext(ctx).With().Info("validating nipost", log.String("expected_challenge_hash", expectedChallengeHash.String()), atx.ID())
 
-	pubKey := signing.NewPublicKey(atx.NodeID().ToBytes())
-	leaves, err := h.nipostValidator.Validate(*pubKey, atx.NIPost, *expectedChallengeHash, uint(atx.NumUnits))
+	commitment, err := h.getCommitmentFromAtx(atx)
 	if err != nil {
-		return nil, fmt.Errorf("invalid nipost: %v", err)
+		return nil, fmt.Errorf("validation failed: initial atx not found: %w", err)
+	}
+
+	leaves, err := h.nipostValidator.Validate(commitment, atx.NIPost, *expectedChallengeHash, uint(atx.NumUnits))
+	if err != nil {
+		return nil, fmt.Errorf("invalid nipost: %w", err)
 	}
 
 	return atx.Verify(baseTickHeight, leaves/h.tickSize)
 }
 
-// ContextuallyValidateAtx ensures that the previous ATX referenced is the last known ATX for the referenced miner ID.
-// If a previous ATX is not referenced, it validates that indeed there's no previous known ATX for that miner ID.
-func (h *Handler) ContextuallyValidateAtx(atx *types.VerifiedActivationTx) error {
-	if atx.PrevATXID != *types.EmptyATXID {
-		lastAtx, err := atxs.GetLastIDByNodeID(h.cdb, atx.NodeID())
+func (h *Handler) validateInitialAtx(ctx context.Context, atx *types.ActivationTx) error {
+	if atx.Sequence != 0 {
+		return fmt.Errorf("no prevATX declared, but sequence number not zero")
+	}
+
+	if atx.InitialPost == nil {
+		return fmt.Errorf("no prevATX declared, but initial Post is not included")
+	}
+
+	if atx.InitialPostIndices == nil {
+		return fmt.Errorf("no prevATX declared, but initial Post indices is not included in challenge")
+	}
+
+	if !bytes.Equal(atx.InitialPost.Indices, atx.InitialPostIndices) {
+		return fmt.Errorf("initial Post indices included in challenge does not equal to the initial Post indices included in the atx")
+	}
+
+	if atx.CommitmentATX == nil {
+		return fmt.Errorf("no prevATX declared, but commitmentATX is missing")
+	}
+
+	if *atx.CommitmentATX == h.goldenATXID && !atx.PublishEpoch().IsGenesis() {
```

Comment:
i am missing something for the commitment ATX logic.

for genesis miners, they can only have golden atx id for commitment atx.
and given that miners don't change commitment atx for the same node id, all their subsequent atxs are going to have golden atx as commitment atx.

Reply:
That's mostly correct. Only the very first ATX published by a node has `atx.CommitmentATX` set and the commitment ATX of a node doesn't change over its lifetime.

Subsequent ATXs don't have a Commitment ATX but the NiPosts in them will use the commitment ATX set in the first ATX of the node.